File: sdks/php/tests/Errors/BlokErrorTest.php
<?php

declare(strict_types=1);

namespace Blok\Blok\Tests\Errors;

use Blok\Blok\Errors\BlokError;
use Blok\Blok\Errors\BlokErrorCategory;
use Blok\Blok\Errors\BlokErrorSeverity;
use Blok\Blok\Errors\BuildContextSnapshot;
use Blok\Blok\Errors\NodeException;
use Blok\Blok\Errors\Origin;
use PHPUnit\Framework\TestCase;

/**
 * Unit tests for the structured {@see BlokError} per master plan §17.
 *
 * Coverage parallels Python (`test_blok_error.py`), Go (`blok_error_test.go`),
 * Rust (`blok_error::tests`), Java (`BlokErrorTest`), C# (`BlokErrorTests`),
 * and Ruby (`blok_error_test.rb`). Each SDK exhaustively tests the same API
 * surface so the cross-language wire shape stays in lockstep.
 */
final class BlokErrorTest extends TestCase
{
    // ===== Category defaults =================================================

    public function test_category_default_status_matches_table(): void
    {
        $this->assertSame(400, BlokErrorCategory::Validation->defaultHttpStatus());
        $this->assertSame(500, BlokErrorCategory::Configuration->defaultHttpStatus());
        $this->assertSame(502, BlokErrorCategory::Dependency->defaultHttpStatus());
        $this->assertSame(504, BlokErrorCategory::Timeout->defaultHttpStatus());
        $this->assertSame(403, BlokErrorCategory::Permission->defaultHttpStatus());
        $this->assertSame(429, BlokErrorCategory::RateLimit->defaultHttpStatus());
        $this->assertSame(404, BlokErrorCategory::NotFound->defaultHttpStatus());
        $this->assertSame(409, BlokErrorCategory::Conflict->defaultHttpStatus());
        $this->assertSame(499, BlokErrorCategory::Cancelled->defaultHttpStatus());
        $this->assertSame(500, BlokErrorCategory::Internal->defaultHttpStatus());
        $this->assertSame(502, BlokErrorCategory::Protocol->defaultHttpStatus());
        $this->assertSame(422, BlokErrorCategory::Data->defaultHttpStatus());
    }

    public function test_category_default_retryable_matches_table(): void
    {
        $this->assertTrue(BlokErrorCategory::Dependency->defaultRetryable());
        $this->assertTrue(BlokErrorCategory::Timeout->defaultRetryable());
        $this->assertTrue(BlokErrorCategory::RateLimit->defaultRetryable());
        $this->assertFalse(BlokErrorCategory::Validation->defaultRetryable());
        $this->assertFalse(BlokErrorCategory::Internal->defaultRetryable());
        $this->assertFalse(BlokErrorCategory::Conflict->defaultRetryable());
    }

    public function test_category_parse_unknown_falls_back_to_internal(): void
    {
        $this->assertSame(BlokErrorCategory::Dependency, BlokErrorCategory::parse('DEPENDENCY'));
        $this->assertSame(BlokErrorCategory::Internal,   BlokErrorCategory::parse('not-a-thing'));
        $this->assertSame(BlokErrorCategory::Internal,   BlokErrorCategory::parse(null));
    }

    public function test_severity_parse_falls_back_to_error(): void
    {
        $this->assertSame(BlokErrorSeverity::Info,  BlokErrorSeverity::parse('INFO'));
        $this->assertSame(BlokErrorSeverity::Error, BlokErrorSeverity::parse('xyz'));
        $this->assertSame(BlokErrorSeverity::Error, BlokErrorSeverity::parse(null));
    }

    // ===== Builder ===========================================================

    public function test_builder_dependency_defaults(): void
    {
        $e = BlokError::dependency()->code('X')->message('y')->build();
        $this->assertSame(BlokErrorCategory::Dependency, $e->category);
        $this->assertSame(502, $e->httpStatus);
        $this->assertTrue($e->retryable);
        $this->assertSame(BlokErrorSeverity::Error, $e->severity);
    }

    public function test_builder_validation_defaults(): void
    {
        $e = BlokError::validation()->code('V')->message('v')->build();
        $this->assertSame(BlokErrorCategory::Validation, $e->category);
        $this->assertSame(400, $e->httpStatus);
        $this->assertFalse($e->retryable);
    }

    public function test_builder_overrides_take_priority(): void
    {
        $e = BlokError::dependency()
            ->httpStatus(599)
            ->retryable(false)
            ->severity(BlokErrorSeverity::Fatal)
            ->build();
        $this->assertSame(599, $e->httpStatus);
        $this->assertFalse($e->retryable);
        $this->assertSame(BlokErrorSeverity::Fatal, $e->severity);
    }

    public function test_builder_retry_after_date_interval_to_ms(): void
    {
        $e = BlokError::rateLimit()->retryAfter(new \DateInterval('PT5S'))->build();
        $this->assertSame(5_000, $e->retryAfterMs);
    }

    public function test_builder_retry_after_ms_direct(): void
    {
        $e = BlokError::timeout()->retryAfterMs(750)->build();
        $this->assertSame(750, $e->retryAfterMs);
    }

    public function test_builder_details_round_trip(): void
    {
        $details = ['issues' => [['path' => ['email']]]];
        $e = BlokError::validation()->details($details)->build();
        $this->assertSame('email', $e->details['issues'][0]['path'][0]);
    }

    public function test_builder_cause_populates_causes_list(): void
    {
        $cause = new \RuntimeException('nope');
        $e = BlokError::dependency()->cause($cause)->build();
        $this->assertNotEmpty($e->causes);
        $this->assertSame('INTERNAL', $e->causes[0]['category']);
        $this->assertSame('nope', $e->causes[0]['message']);
    }

    public function test_builder_apply_origin_fills_only_missing(): void
    {
        $origin = Origin::defaults('my-node', '1.2.3');
        $e = BlokError::dependency()->sdk('custom')->applyOrigin($origin)->build();
        $this->assertSame('custom',       $e->sdk);          // explicit value preserved
        $this->assertSame('my-node',      $e->node);         // empty filled
        $this->assertSame('1.2.3',        $e->sdkVersion);
        $this->assertSame('runtime.php',  $e->runtimeKind);
    }

    public function test_all_twelve_category_factories_produce_correct_category(): void
    {
        $this->assertSame(BlokErrorCategory::Validation,    BlokError::validation()->build()->category);
        $this->assertSame(BlokErrorCategory::Configuration, BlokError::configuration()->build()->category);
        $this->assertSame(BlokErrorCategory::Dependency,    BlokError::dependency()->build()->category);
        $this->assertSame(BlokErrorCategory::Timeout,       BlokError::timeout()->build()->category);
        $this->assertSame(BlokErrorCategory::Permission,    BlokError::permission()->build()->category);
        $this->assertSame(BlokErrorCategory::RateLimit,     BlokError::rateLimit()->build()->category);
        $this->assertSame(BlokErrorCategory::NotFound,      BlokError::notFound()->build()->category);
        $this->assertSame(BlokErrorCategory::Conflict,      BlokError::conflict()->build()->category);
        $this->assertSame(BlokErrorCategory::Cancelled,     BlokError::cancelled()->build()->category);
        $this->assertSame(BlokErrorCategory::Internal,      BlokError::internal()->build()->category);
        $this->assertSame(BlokErrorCategory::Protocol,      BlokError::protocol()->build()->category);
        $this->assertSame(BlokErrorCategory::Data,          BlokError::data()->build()->category);
    }

    public function test_of_produces_generic_factory(): void
    {
        $e = BlokError::of(BlokErrorCategory::Data)->code('x')->message('y')->build();
        $this->assertSame(BlokErrorCategory::Data, $e->category);
        $this->assertSame(422, $e->httpStatus);
    }

    // ===== fromUnknown =======================================================

    public function test_from_unknown_passes_through_typed_blok_error(): void
    {
        $origin = Origin::defaults('auto-node', '1.2.3');
        $original = BlokError::rateLimit()->code('UPSTREAM_RATE_LIMITED')->message('limit hit')->build();
        $recovered = BlokError::fromUnknown($original, $origin);
        $this->assertSame($original, $recovered);
        $this->assertSame('auto-node', $recovered->node);
        $this->assertSame('1.2.3',     $recovered->sdkVersion);
        $this->assertSame(BlokErrorCategory::RateLimit, $recovered->category);
    }

    public function test_from_unknown_wraps_throwable(): void
    {
        $origin = Origin::defaults('auto', '1.0.0');
        $cause = new \RuntimeException('disk full');
        $wrapped = BlokError::fromUnknown($cause, $origin);
        $this->assertSame(BlokErrorCategory::Internal, $wrapped->category);
        $this->assertSame('disk full', $wrapped->getMessage());
        $this->assertStringStartsWith('UNCAUGHT_', $wrapped->errorCode);
    }

    public function test_from_unknown_wraps_string(): void
    {
        $wrapped = BlokError::fromUnknown('boom', Origin::defaults('x', '1.0.0'));
        $this->assertSame(BlokErrorCategory::Internal, $wrapped->category);
        $this->assertSame('boom', $wrapped->getMessage());
        $this->assertSame('UNCAUGHT_ERROR', $wrapped->errorCode);
        $this->assertSame('boom', $wrapped->details['message']);
    }

    public function test_from_unknown_wraps_array(): void
    {
        $raw = ['message' => 'from-map', 'custom' => 42];
        $wrapped = BlokError::fromUnknown($raw, Origin::defaults('x', '1.0.0'));
        $this->assertSame('from-map', $wrapped->getMessage());
        $this->assertSame(BlokErrorCategory::Internal, $wrapped->category);
        $this->assertSame(42, $wrapped->details['custom']);
    }

    public function test_from_unknown_handles_null(): void
    {
        $wrapped = BlokError::fromUnknown(null, Origin::defaults('x', '1.0.0'));
        $this->assertSame('node error', $wrapped->getMessage());
        $this->assertSame(BlokErrorCategory::Internal, $wrapped->category);
    }

    public function test_from_unknown_wraps_legacy_node_exception(): void
    {
        $legacy = NodeException::network('postgres unreachable');
        $wrapped = BlokError::fromUnknown($legacy, Origin::defaults('x', '1.0.0'));
        $this->assertSame(BlokErrorCategory::Internal, $wrapped->category);
        $this->assertSame('UNCAUGHT_NODEEXCEPTION', $wrapped->errorCode);
        $this->assertStringContainsString('postgres unreachable', $wrapped->getMessage());
        $this->assertNotNull($wrapped->details);
    }

    // ===== toArray / fromArray ===============================================

    public function test_to_array_and_from_array_round_trip(): void
    {
        $details = ['a' => 1];
        $e = BlokError::dependency()
            ->code('CODE')
            ->message('msg')
            ->description('desc')
            ->remediation('rem')
            ->docUrl('https://example.com')
            ->retryable(true)
            ->retryAfterMs(1234)
            ->details($details)
            ->node('n')
            ->sdk('blok-php')
            ->sdkVersion('1.0.0')
            ->runtimeKind('runtime.php')
            ->build();

        $array = $e->toArray();
        $this->assertSame('DEPENDENCY', $array['category']);
        $this->assertSame('CODE',       $array['code']);
        $this->assertSame(502,          $array['http_status']);
        $this->assertSame(1234,         $array['retry_after_ms']);

        $restored = BlokError::fromArray($array);
        $this->assertSame(BlokErrorCategory::Dependency, $restored->category);
        $this->assertSame('CODE',                $restored->errorCode);
        $this->assertSame('msg',                 $restored->getMessage());
        $this->assertSame('desc',                $restored->description);
        $this->assertSame(1234,                  $restored->retryAfterMs);
        $this->assertSame('https://example.com', $restored->docUrl);
    }

    public function test_from_array_accepts_camel_case_keys(): void
    {
        $raw = [
            'category'     => 'RATE_LIMIT',
            'severity'     => 'ERROR',
            'code'         => 'RL',
            'message'      => 'too many',
            'httpStatus'   => 429,
            'retryable'    => true,
            'retryAfterMs' => 60_000,
            'at'           => '2026-04-29T00:00:00+00:00',
            'sdkVersion'   => '1.0.0',
            'runtimeKind'  => 'runtime.php',
            'docUrl'       => 'https://docs/example',
        ];
        $e = BlokError::fromArray($raw);
        $this->assertSame(BlokErrorCategory::RateLimit, $e->category);
        $this->assertSame(429,                  $e->httpStatus);
        $this->assertSame(60_000,               $e->retryAfterMs);
        $this->assertSame('1.0.0',              $e->sdkVersion);
        $this->assertSame('runtime.php',        $e->runtimeKind);
        $this->assertSame('https://docs/example', $e->docUrl);
    }

    public function test_from_array_accepts_causes_list(): void
    {
        $raw = [
            'category' => 'DEPENDENCY',
            'severity' => 'ERROR',
            'code'     => 'X',
            'message'  => 'y',
            'causes'   => [
                ['message' => 'inner', 'category' => 'INTERNAL'],
            ],
        ];
        $e = BlokError::fromArray($raw);
        $this->assertCount(1, $e->causes);
        $this->assertSame('inner', $e->causes[0]['message']);
    }

    // ===== Display / Exception semantics =====================================

    public function test_to_string_formats_category_and_message(): void
    {
        $e = BlokError::dependency()->code('X')->message('nope')->build();
        $this->assertSame('[DEPENDENCY] nope', (string) $e);
    }

    public function test_can_be_thrown_as_exception(): void
    {
        $e = BlokError::timeout()->code('X')->message('y')->build();
        $this->expectException(BlokError::class);
        throw $e;
    }

    // ===== uncaught code derivation ==========================================

    public function test_uncaught_code_strips_namespace_and_uppercases(): void
    {
        $this->assertSame('UNCAUGHT_RUNTIMEEXCEPTION', BlokError::uncaughtCode(\RuntimeException::class));
        $this->assertSame('UNCAUGHT_BLOKERROR',         BlokError::uncaughtCode(BlokError::class));
        $this->assertSame('UNCAUGHT_ERROR',             BlokError::uncaughtCode(null));
        $this->assertSame('UNCAUGHT_ERROR',             BlokError::uncaughtCode(''));
    }

    // ===== cause-chain flattening ============================================

    public function test_flatten_causes_walks_get_previous_chain(): void
    {
        $inner = new \RuntimeException('inner');
        $wrap = new \LogicException('wrapped', 0, $inner);
        $causes = BlokError::flattenCauses($wrap);
        $this->assertCount(2, $causes);
        $this->assertSame('wrapped', $causes[0]['message']);
        $this->assertSame('inner',   $causes[1]['message']);
    }

    public function test_flatten_causes_lifts_blok_error_link(): void
    {
        $inner = BlokError::notFound()->code('INNER')->message('inner-msg')->build();
        $causes = BlokError::flattenCauses($inner);
        $this->assertSame('INNER',     $causes[0]['code']);
        $this->assertSame('NOT_FOUND', $causes[0]['category']);
    }

    // ===== BuildContextSnapshot ==============================================

    public function test_snapshot_preserves_small_payload(): void
    {
        $inputs = ['a' => 1];
        $vars = ['k1' => 'v1'];
        $snap = BuildContextSnapshot::of($inputs, $vars);
        $this->assertSame(1,    $snap['inputs']['a']);
        $this->assertSame('v1', $snap['vars']['k1']);
    }

    public function test_snapshot_caps_at_max_bytes(): void
    {
        $inputs = [];
        $vars = [];
        $filler = str_repeat('x', 100);
        for ($i = 0; $i < 80; $i++) {
            $vars[sprintf('k%03d', $i)] = $filler;
        }
        $snap = BuildContextSnapshot::of($inputs, $vars);
        $bytes = strlen(json_encode($snap, JSON_UNESCAPED_SLASHES) ?: '');
        $this->assertLessThanOrEqual(BlokError::CONTEXT_SNAPSHOT_MAX_BYTES + 64, $bytes);
    }

    public function test_snapshot_keeps_last_n_keys(): void
    {
        $inputs = [];
        $vars = [];
        for ($i = 0; $i < 32; $i++) {
            $vars[sprintf('k%02d', $i)] = $i;
        }
        $snap = BuildContextSnapshot::withOpts($inputs, $vars, 0, 5);
        $kept = $snap['vars'];
        $this->assertCount(5, $kept);
        $this->assertArrayHasKey('k31', $kept);
        $this->assertArrayNotHasKey('k00', $kept);
    }

    public function test_snapshot_disables_var_keys_when_zero(): void
    {
        $inputs = [];
        $vars = ['only' => 1];
        $snap = BuildContextSnapshot::withOpts($inputs, $vars, 0, 0);
        $this->assertEmpty($snap['vars']);
    }

    // ===== Origin ============================================================

    public function test_origin_defaults_uses_sdk_constants(): void
    {
        $o = Origin::defaults('n', '1.2.3');
        $this->assertSame(BlokError::DEFAULT_SDK_NAME,     $o->sdk);
        $this->assertSame(BlokError::DEFAULT_RUNTIME_KIND, $o->runtimeKind);
        $this->assertSame('n',     $o->node);
        $this->assertSame('1.2.3', $o->sdkVersion);
    }

    public function test_apply_origin_if_missing_preserves_explicit_fields(): void
    {
        $e = BlokError::internal()->code('X')->message('y')->node('explicit')->build();
        $e->applyOriginIfMissing(Origin::defaults('auto', '1.0.0'));
        $this->assertSame('explicit',                    $e->node);
        $this->assertSame(BlokError::DEFAULT_SDK_NAME,   $e->sdk);
        $this->assertSame(BlokError::DEFAULT_RUNTIME_KIND, $e->runtimeKind);
    }
}

File: sdks/ruby/test/errors/blok_error_test.rb
# frozen_string_literal: true

require_relative "../test_helper"
require "json"

# Unit tests for the structured +Blok::Errors::BlokError+ per master plan §17.
#
# Coverage parallels Python (+test_blok_error.py+), Go (+blok_error_test.go+),
# Rust (+blok_error::tests+), Java (+BlokErrorTest+), and C#
# (+BlokErrorTests+). Each SDK exhaustively tests the same API surface so
# the cross-language wire shape stays in lockstep.
class TestBlokError < Minitest::Test
  C  = Blok::Errors::BlokError::Category
  S  = Blok::Errors::BlokError::Severity
  BE = Blok::Errors::BlokError

  # ===== Category defaults =================================================

  def test_category_default_status_matches_table
    assert_equal 400, C::DEFAULT_HTTP_STATUS[C::VALIDATION]
    assert_equal 500, C::DEFAULT_HTTP_STATUS[C::CONFIGURATION]
    assert_equal 502, C::DEFAULT_HTTP_STATUS[C::DEPENDENCY]
    assert_equal 504, C::DEFAULT_HTTP_STATUS[C::TIMEOUT]
    assert_equal 403, C::DEFAULT_HTTP_STATUS[C::PERMISSION]
    assert_equal 429, C::DEFAULT_HTTP_STATUS[C::RATE_LIMIT]
    assert_equal 404, C::DEFAULT_HTTP_STATUS[C::NOT_FOUND]
    assert_equal 409, C::DEFAULT_HTTP_STATUS[C::CONFLICT]
    assert_equal 499, C::DEFAULT_HTTP_STATUS[C::CANCELLED]
    assert_equal 500, C::DEFAULT_HTTP_STATUS[C::INTERNAL]
    assert_equal 502, C::DEFAULT_HTTP_STATUS[C::PROTOCOL]
    assert_equal 422, C::DEFAULT_HTTP_STATUS[C::DATA]
  end

  def test_category_default_retryable_matches_table
    assert C::DEFAULT_RETRYABLE[C::DEPENDENCY]
    assert C::DEFAULT_RETRYABLE[C::TIMEOUT]
    assert C::DEFAULT_RETRYABLE[C::RATE_LIMIT]
    refute C::DEFAULT_RETRYABLE[C::VALIDATION]
    refute C::DEFAULT_RETRYABLE[C::INTERNAL]
    refute C::DEFAULT_RETRYABLE[C::CONFLICT]
  end

  def test_category_parse_unknown_falls_back_to_internal
    assert_equal C::DEPENDENCY, C.parse("DEPENDENCY")
    assert_equal C::INTERNAL,   C.parse("not-a-thing")
    assert_equal C::INTERNAL,   C.parse(nil)
  end

  def test_severity_parse_falls_back_to_error
    assert_equal S::INFO,  S.parse("INFO")
    assert_equal S::ERROR, S.parse("xyz")
    assert_equal S::ERROR, S.parse(nil)
  end

  # ===== Builder ===========================================================

  def test_builder_dependency_defaults
    e = BE.dependency(code: "X", message: "y")
    assert_equal C::DEPENDENCY, e.category
    assert_equal 502, e.http_status
    assert e.retryable
    assert_equal S::ERROR, e.severity
  end

  def test_builder_validation_defaults
    e = BE.validation(code: "V", message: "v")
    assert_equal C::VALIDATION, e.category
    assert_equal 400, e.http_status
    refute e.retryable
  end

  def test_builder_overrides_take_priority
    e = BE.dependency(code: "X", message: "y", http_status: 599, retryable: false, severity: S::FATAL)
    assert_equal 599, e.http_status
    refute e.retryable
    assert_equal S::FATAL, e.severity
  end

  def test_builder_retry_after_ms_direct
    e = BE.timeout(code: "T", message: "t", retry_after_ms: 750)
    assert_equal 750, e.retry_after_ms
  end

  def test_builder_details_round_trip
    details = { "issues" => [{ "path" => ["email"] }] }
    e = BE.validation(code: "V", message: "m", details: details)
    assert_equal "email", e.details["issues"][0]["path"][0]
  end

  def test_builder_cause_populates_causes_list
    cause = StandardError.new("nope")
    e = BE.dependency(code: "X", message: "y", cause: cause)
    refute_empty e.causes
    assert_equal "INTERNAL", e.causes[0]["category"]
    assert_equal "nope", e.causes[0]["message"]
  end

  def test_all_twelve_category_factories_produce_correct_category
    assert_equal C::VALIDATION,    BE.validation(code: "x", message: "y").category
    assert_equal C::CONFIGURATION, BE.configuration(code: "x", message: "y").category
    assert_equal C::DEPENDENCY,    BE.dependency(code: "x", message: "y").category
    assert_equal C::TIMEOUT,       BE.timeout(code: "x", message: "y").category
    assert_equal C::PERMISSION,    BE.permission(code: "x", message: "y").category
    assert_equal C::RATE_LIMIT,    BE.rate_limit(code: "x", message: "y").category
    assert_equal C::NOT_FOUND,     BE.not_found(code: "x", message: "y").category
    assert_equal C::CONFLICT,      BE.conflict(code: "x", message: "y").category
    assert_equal C::CANCELLED,     BE.cancelled(code: "x", message: "y").category
    assert_equal C::INTERNAL,      BE.internal(code: "x", message: "y").category
    assert_equal C::PROTOCOL,      BE.protocol(code: "x", message: "y").category
    assert_equal C::DATA,          BE.data(code: "x", message: "y").category
  end

  # ===== from_unknown ======================================================

  def test_from_unknown_passes_through_typed_blok_error
    origin = BE::Origin.defaults(node: "auto-node", sdk_version: "1.2.3")
    original = BE.rate_limit(code: "UPSTREAM_RATE_LIMITED", message: "limit hit")
    recovered = BE.from_unknown(original, origin: origin)
    assert_same original, recovered
    assert_equal "auto-node", recovered.node
    assert_equal "1.2.3", recovered.sdk_version
    assert_equal C::RATE_LIMIT, recovered.category
  end

  def test_from_unknown_wraps_throwable
    origin = BE::Origin.defaults(node: "auto", sdk_version: "1.0.0")
    cause = ArgumentError.new("disk full")
    wrapped = BE.from_unknown(cause, origin: origin)
    assert_equal C::INTERNAL, wrapped.category
    assert_equal "disk full", wrapped.message
    assert wrapped.code.start_with?("UNCAUGHT_")
  end

  def test_from_unknown_wraps_string
    origin = BE::Origin.defaults(node: "x", sdk_version: "1.0.0")
    wrapped = BE.from_unknown("boom", origin: origin)
    assert_equal C::INTERNAL, wrapped.category
    assert_equal "boom", wrapped.message
    assert_equal "UNCAUGHT_ERROR", wrapped.code
    assert_equal "boom", wrapped.details["message"]
  end

  def test_from_unknown_wraps_hash
    origin = BE::Origin.defaults(node: "x", sdk_version: "1.0.0")
    raw = { "message" => "from-map", "custom" => 42 }
    wrapped = BE.from_unknown(raw, origin: origin)
    assert_equal "from-map", wrapped.message
    assert_equal C::INTERNAL, wrapped.category
    assert_equal 42, wrapped.details["custom"]
  end

  def test_from_unknown_handles_nil
    origin = BE::Origin.defaults(node: "x", sdk_version: "1.0.0")
    wrapped = BE.from_unknown(nil, origin: origin)
    assert_equal "node error", wrapped.message
    assert_equal C::INTERNAL, wrapped.category
  end

  def test_from_unknown_wraps_legacy_node_error
    origin = BE::Origin.defaults(node: "x", sdk_version: "1.0.0")
    legacy = Blok::Errors::NodeError.network("postgres unreachable")
    wrapped = BE.from_unknown(legacy, origin: origin)
    assert_equal C::INTERNAL, wrapped.category
    assert_equal "UNCAUGHT_NODEERROR", wrapped.code
    assert_includes wrapped.message, "postgres unreachable"
    refute_nil wrapped.details
  end

  # ===== to_hash / from_hash ===============================================

  def test_to_hash_and_from_hash_round_trip
    details = { "a" => 1 }
    e = BE.dependency(
      code: "CODE", message: "msg", description: "desc", remediation: "rem",
      doc_url: "https://example.com", retryable: true, retry_after_ms: 1234,
      details: details, node: "n", sdk: "blok-ruby", sdk_version: "1.0.0",
      runtime_kind: "runtime.ruby"
    )
    map = e.to_hash
    assert_equal "DEPENDENCY", map["category"]
    assert_equal "CODE", map["code"]
    assert_equal 502, map["http_status"]
    assert_equal 1234, map["retry_after_ms"]

    restored = BE.from_hash(map)
    assert_equal C::DEPENDENCY, restored.category
    assert_equal "CODE", restored.code
    assert_equal "msg", restored.message
    assert_equal "desc", restored.description
    assert_equal 1234, restored.retry_after_ms
    assert_equal "https://example.com", restored.doc_url
  end

  def test_from_hash_accepts_camel_case_keys
    raw = {
      "category"     => "RATE_LIMIT",
      "severity"     => "ERROR",
      "code"         => "RL",
      "message"      => "too many",
      "httpStatus"   => 429,
      "retryable"    => true,
      "retryAfterMs" => 60_000,
      "at"           => "2026-04-29T00:00:00Z",
      "sdkVersion"   => "1.0.0",
      "runtimeKind"  => "runtime.ruby",
      "docUrl"       => "https://docs/example"
    }
    e = BE.from_hash(raw)
    assert_equal C::RATE_LIMIT, e.category
    assert_equal 429, e.http_status
    assert_equal 60_000, e.retry_after_ms
    assert_equal "1.0.0", e.sdk_version
    assert_equal "runtime.ruby", e.runtime_kind
    assert_equal "https://docs/example", e.doc_url
  end

  def test_from_hash_accepts_causes_list
    raw = {
      "category" => "DEPENDENCY",
      "severity" => "ERROR",
      "code"     => "X",
      "message"  => "y",
      "causes"   => [
        { "message" => "inner", "category" => "INTERNAL" }
      ]
    }
    e = BE.from_hash(raw)
    assert_equal 1, e.causes.length
    assert_equal "inner", e.causes[0]["message"]
  end

  # ===== Display / Exception semantics =====================================

  def test_to_string_formats_category_and_message
    e = BE.dependency(code: "X", message: "nope")
    assert_equal "[DEPENDENCY] nope", e.to_s
  end

  def test_can_be_raised_as_standard_error
    e = BE.timeout(code: "X", message: "y")
    assert_raises(BE) { raise e }
  end

  # ===== uncaught code derivation ==========================================

  def test_uncaught_code_strips_namespace_and_uppercases
    assert_equal "UNCAUGHT_IOERROR", BE._uncaught_code(IOError)
    assert_equal "UNCAUGHT_BLOKERROR", BE._uncaught_code(BE)
    assert_equal "UNCAUGHT_ERROR", BE._uncaught_code(nil)
  end

  # ===== cause-chain flattening ============================================

  def test_flatten_causes_walks_cause_chain
    inner = nil
    wrap = nil
    begin
      begin
        raise IOError, "inner"
      rescue IOError
        raise StandardError, "wrapped"
      end
    rescue StandardError => e
      wrap = e
    end
    causes = BE.flatten_causes(wrap)
    assert_equal 2, causes.length
    assert_equal "wrapped", causes[0]["message"]
    assert_equal "inner",   causes[1]["message"]
  end

  def test_flatten_causes_lifts_blok_error_link
    inner = BE.not_found(code: "INNER", message: "inner-msg")
    causes = BE.flatten_causes(inner)
    assert_equal "INNER",     causes[0]["code"]
    assert_equal "NOT_FOUND", causes[0]["category"]
  end

  # ===== BuildContextSnapshot ==============================================

  def test_snapshot_preserves_small_payload
    inputs = { "a" => 1 }
    vars = { "k1" => "v1" }
    snap = Blok::Errors::BuildContextSnapshot.of(inputs: inputs, vars: vars)
    assert_equal 1,    snap["inputs"]["a"]
    assert_equal "v1", snap["vars"]["k1"]
  end

  def test_snapshot_caps_at_max_bytes
    inputs = {}
    vars = {}
    filler = "x" * 100
    80.times { |i| vars[format("k%03d", i)] = filler }
    snap = Blok::Errors::BuildContextSnapshot.of(inputs: inputs, vars: vars)
    bytes = JSON.generate(snap).bytesize
    assert bytes <= BE::CONTEXT_SNAPSHOT_MAX_BYTES + 64,
           "snapshot #{bytes} bytes exceeded budget #{BE::CONTEXT_SNAPSHOT_MAX_BYTES}"
  end

  def test_snapshot_keeps_last_n_keys
    inputs = {}
    vars = {}
    32.times { |i| vars[format("k%02d", i)] = i }
    snap = Blok::Errors::BuildContextSnapshot.with_opts(
      inputs: inputs, vars: vars, max_bytes: 0, max_vars_keys: 5
    )
    kept = snap["vars"]
    assert_equal 5, kept.length
    assert kept.key?("k31")
    refute kept.key?("k00")
  end

  def test_snapshot_disables_var_keys_when_zero
    inputs = {}
    vars = { "only" => 1 }
    snap = Blok::Errors::BuildContextSnapshot.with_opts(
      inputs: inputs, vars: vars, max_bytes: 0, max_vars_keys: 0
    )
    assert_empty snap["vars"]
  end

  # ===== Origin ============================================================

  def test_origin_defaults_uses_sdk_constants
    o = BE::Origin.defaults(node: "n", sdk_version: "1.2.3")
    assert_equal BE::DEFAULT_SDK_NAME,     o.sdk
    assert_equal BE::DEFAULT_RUNTIME_KIND, o.runtime_kind
    assert_equal "n",     o.node
    assert_equal "1.2.3", o.sdk_version
  end

  def test_apply_origin_if_missing_preserves_explicit_fields
    e = BE.internal(code: "X", message: "y", node: "explicit")
    e.apply_origin_if_missing(BE::Origin.defaults(node: "auto", sdk_version: "1.0.0"))
    assert_equal "explicit",                e.node
    assert_equal BE::DEFAULT_SDK_NAME,      e.sdk
    assert_equal BE::DEFAULT_RUNTIME_KIND,  e.runtime_kind
  end
end
